### PR TITLE
registrant-extra-info: Pass consistent error and optional change handler props (3)

### PR DESCRIPTION
**Pull request by nbloomf**
#### Changes proposed in this Pull Request

* Part 2 of #39837. Composite checkout uses calypso's domain contact forms as managed components, so it needs to be able to pass errors and a custom change event handler. This PR adds these to the extra contact field components for specific TLDs - ca, fr, uk. (The readme indicates this is the intended behavior anyway, but afaict it was not consistently implemented.)
* All TLD specific forms now take a `contactDetailsValidationErrors` prop. This changes the name of the `validationErrors` prop taken by the UK form, which was only explicitly passed in tests.

#### Testing instructions

This patch should have no observable effects yet. Test that this is the case by trying to get through the contact details pages in _regular_ checkout, paying attention to the following cases:

1. Have a .ca, .uk, or .fr domain in your cart. Verify that the extra fields are shown and that there are no JS errors thrown by the contact form components.
2. Try to enter bad data in the extra fields and verify that validation errors are shown.

### CODE REVIEW

File: client/components/domains/registrant-extra-info/ca-form.jsx [hunk 1]
```diff
@@ -35,6 +35,8 @@ export class RegistrantExtraInfoCaForm extends React.PureComponent {
 	static propTypes = {
 		contactDetails: PropTypes.object.isRequired,
 		ccTldDetails: PropTypes.object.isRequired,
+		onContactDetailsChange: PropTypes.func,
```
Comment (sirbrillig) [inline client/components/domains/registrant-extra-info/ca-form.jsx]: Not really related to your PR but it might be worth filling in the missing `updateContactDetailsCache` (required) proptype here too.

File: client/components/domains/registrant-extra-info/ca-form.jsx [hunk 2]
```diff
@@ -154,7 +160,9 @@ export class RegistrantExtraInfoCaForm extends React.PureComponent {
 	}
 
 	getOrganizationErrorMessage() {
-		let message = ( this.state.errorMessages.organization || [] ).join( '\n' );
+		let message = this.props?.contactDetailsValidationErrors?.organization
```
Comment (sirbrillig) [inline client/components/domains/registrant-extra-info/ca-form.jsx]: I _think_ this could safely be

```js
let message = this.props.contactDetailsValidationErrors?.organization || ( this.state.errorMessages.organization || [] ).join( '\n' );
```

Also notice that `this.props` will always exist so there's no need to put a question mark after it (the `?.` operator applies to the property just before it). `this.props?.contactDetailsValidationErrors` is equivalent to `this.props && this.props.contactDetailsValidationErrors`.

File: client/components/domains/registrant-extra-info/fr-form.jsx [hunk 3]
```diff
@@ -83,17 +84,22 @@ class RegistrantExtraInfoFrForm extends React.PureComponent {
...
-		this.props.updateContactDetailsCache( {
+		const payload = {
 			extra: {
 				fr: { registrantType: defaultRegistrantType },
 			},
-		} );
+		};
+
+		this.props.updateContactDetailsCache( payload );
+		this.props?.onContactDetailsChange( payload );
```
Comment (sirbrillig) [inline client/components/domains/registrant-extra-info/fr-form.jsx]: You probably meant `this.props.onContactDetailsChange?.(payload)` since `this.props` will always exist.

File: client/components/domains/registrant-extra-info/fr-form.jsx [hunk 4]
```diff
@@ -83,17 +84,22 @@ class RegistrantExtraInfoFrForm extends React.PureComponent {
...
 	}
 
 	updateContactDetails( field, value ) {
 		const sanitizedValue = this.sanitizeField( field, value );
 		debug( 'Setting ' + field + ' to ' + value );
-		this.props.updateContactDetailsCache( set( {}, field, sanitizedValue ) );
+		const payload = set( {}, field, sanitizedValue );
+		this.props.updateContactDetailsCache( payload );
+		this.props.onContactDetailsChange( payload );
```
Comment (sirbrillig) [inline client/components/domains/registrant-extra-info/fr-form.jsx]: This should probably be `this.props.onContactDetailsChange?.(payload)` in case `onContactDetailsChange` is not provided.

File: client/components/domains/registrant-extra-info/fr-form.jsx [hunk 5]
```diff
@@ -83,17 +84,22 @@ class RegistrantExtraInfoFrForm extends React.PureComponent {
...
+		this.props.updateContactDetailsCache( payload );
+		this.props?.onContactDetailsChange( payload );
 	}
 
 	updateContactDetails( field, value ) {
 		const sanitizedValue = this.sanitizeField( field, value );
 		debug( 'Setting ' + field + ' to ' + value );
-		this.props.updateContactDetailsCache( set( {}, field, sanitizedValue ) );
+		const payload = set( {}, field, sanitizedValue );
+		this.props.updateContactDetailsCache( payload );
```
Comment (sirbrillig) [inline client/components/domains/registrant-extra-info/fr-form.jsx]: Interestingly, `updateContactDetailsCache` is not marked as required in the proptypes, and it has no default, so this (and every instance of that function in this class) should be `this.props.updateContactDetailsCache?.(payload)`. Alternatively, the proptype should be changed to make it required since that's the current behavior.

I know that's not part of your PR but it's a small fix for an existing bug and I'd say it's worthwhile.

File: client/components/domains/registrant-extra-info/uk-form.jsx [hunk 6]
```diff
@@ -30,6 +30,8 @@ export class RegistrantExtraInfoUkForm extends React.PureComponent {
 	static propTypes = {
 		contactDetails: PropTypes.object.isRequired,
 		ccTldDetails: PropTypes.object.isRequired,
+		onContactDetailsChange: PropTypes.func,
+		contactDetailsValidationErrors: PropTypes.object,
```
Comment (sirbrillig) [inline client/components/domains/registrant-extra-info/uk-form.jsx]: Same as the other files, I'd suggest adding `updateContactDetailsCache` as a required proptype also as long as you're adjacent to it.

File: client/components/domains/registrant-extra-info/uk-form.jsx [hunk 7]
```diff
@@ -75,19 +77,24 @@ export class RegistrantExtraInfoUkForm extends React.PureComponent {
...
-		this.props.updateContactDetailsCache( {
+		const payload = {
 			extra: {
 				uk: pick( defaultValues, neededRequiredDetails ),
 			},
-		} );
+		};
+
+		this.props.updateContactDetailsCache( payload );
+		this.props?.onContactDetailsChange( payload );
```
Comment (sirbrillig) [inline client/components/domains/registrant-extra-info/uk-form.jsx]: Same as before, `this.props` will exist; you probably meant to mark `onContactDetailsChange` as optional here.

File: client/components/domains/registrant-extra-info/uk-form.jsx [hunk 8]
```diff
@@ -75,19 +77,24 @@ export class RegistrantExtraInfoUkForm extends React.PureComponent {
...
 	handleChangeEvent = event => {
-		this.props.updateContactDetailsCache( {
+		const payload = {
 			extra: {
 				uk: { [ camelCase( event.target.id ) ]: event.target.value },
 			},
-		} );
+		};
+		this.props.updateContactDetailsCache( payload );
+		this.props.onContactDetailsChange( payload );
```
Comment (sirbrillig) [inline client/components/domains/registrant-extra-info/uk-form.jsx]: This should probably be `this.props. onContactDetailsChange?.(payload)`.

File: client/components/domains/registrant-extra-info/uk-form.jsx [hunk 9]
```diff
@@ -137,7 +144,7 @@ export class RegistrantExtraInfoUkForm extends React.PureComponent {
 		const { ccTldDetails, translate } = this.props;
 		const registrationNumber = get( ccTldDetails, 'registrationNumber', '' );
 		const registrationNumberErrors = get(
-			this.props.validationErrors,
```
Comment (sirbrillig) [inline client/components/domains/registrant-extra-info/uk-form.jsx]: I'm assuming you've verified that `validationErrors` is never passed to this component? This is the kind of thing that's probably worth mentioning in the commit message as context since the change isn't explicitly part of what the commit title describes.

File: client/components/domains/registrant-extra-info/uk-form.jsx [hunk 10]
```diff
@@ -191,7 +198,10 @@ export class RegistrantExtraInfoUkForm extends React.PureComponent {
 			this.isTradingNameRequired( registrantType ) && 'tradingName',
 			this.isRegistrationNumberRequired( registrantType ) && 'registrationNumber',
 		] );
-		const relevantErrors = pick( get( validationErrors, 'extra.uk', {} ), relevantExtraFields );
+		const relevantErrors = pick(
+			get( contactDetailsValidationErrors, 'extra.uk', {} ),
```
Comment (sirbrillig) [inline client/components/domains/registrant-extra-info/uk-form.jsx]: I'm not sure this is right, but it's hard to parse. Statements like this are why I try to avoid lodash when possible, especially now that JS itself has so many convenient tools that are easier to understand without reading docs.

If I turn this into pure JS, I think it becomes this:

```js
const relevantErrors = Object.keys( this.props.contactDetailsValidationErrors?.extra?.uk ?? {} )
  .reduce( ( errorsObject, errorKey ) => relevantExtraFields.includes( errorKey ) ? { ...errorsObject, [ errorKey ]: this.props.contactDetailsValidationErrors.extra.uk[ errorKey ] } : errorsObject, {} );
const isValid = Object.keys( relevantErrors ).length === 0;
```

And that sure is a lot of work to do just to set `isValid`. It seems to me that this would work just as well:

```js
const isValid = Object.keys( this.props.contactDetailsValidationErrors?.extra?.uk ?? {} )
  .every( errorKey => ! relevantExtraFields.includes( errorKey ) );
```

Furthermore, from what I know of how you're planning to use `contactDetailsValidationErrors`, is an object in `contactDetailsValidationErrors.extra.uk` really what you meant to do?

File: client/components/domains/registrant-extra-info/uk-form.jsx (hunk 9 above)
Comment (nbloomf) [reply client/components/domains/registrant-extra-info/uk-form.jsx]: As far as I can tell no consumers of this component actually pass the `validationErrors` prop except in tests. I checked this by looking for any files that import the default export from `registrant-extra-info`. I have updated the PR description to make this more clear -- thanks for catching that.

File: client/components/domains/registrant-extra-info/uk-form.jsx (hunk 10 above)
Comment (nbloomf) [reply client/components/domains/registrant-extra-info/uk-form.jsx]: For what it's worth, that line only appears in this diff because of the `validationErrors -> contactDetailsValidationErrors` renaming. I am totally ok with getting rid of some lodash where we can though. :)

I think what's happening here is that which fields are required in this form depends on the selected registrant type, and the component is doing some logic to validate this on the frontend. I think your rewrite is correct.

As for your last question, if I understand correctly, the answer is yes; if there are validation errors in these fields, the object returned by the endpoint has an object at `extra.uk` with a property for each errored field.

Comment (sirbrillig) [reply client/components/domains/registrant-extra-info/uk-form.jsx]: Ok! If the data is correct, that's fine. Now that I re-read my comment it sounds a little like I was saying "how dare you use lodash" but I really meant "since this existing code uses lodash, it's hard for me to parse the effects of your change". I'll leave the decision about what to do here to you. 😄

Comment (nbloomf) [reply client/components/domains/registrant-extra-info/uk-form.jsx]: Haha- no worries. I also find lodash to be confusing, having come to JS after the era in which it was necessary. A lot of times I end up leaving existing code alone even when it probably could be made more clear because I'm not confident I won't break something.

**Review by sirbrillig — APPROVED** [review]
Looks good to me. Testing works as described.
